src/value.rs

```diff
@@ -115,12 +117,39 @@ impl<'de> Deserializer<'de> for Parser<'de> {
                 SeqDeserializer::new(std::iter::once(self)).deserialize_seq(visitor)
             }
             Value::Map(values) => {
-                let values = values.into_iter().map(|(_, val)| Self {
-                    current: val,
-                    config: self.config,
-                });
+                // Convert the map into a sequence
+                fn converter<'de, I, K>(
+                    config: &'de Config,
+                    iter: I,
+                ) -> impl Iterator<Item = Parser<'de>>
+                where
+                    I: IntoIterator<Item = (K, Value)>,
+                {
+                    iter.into_iter().map(|(_, val)| Parser {
+                        current: val,
+                        config,
+                    })
+                }
 
-                SeqDeserializer::new(values).deserialize_seq(visitor)
+                // Check if we can sort these keys numerically.
+                //
+                // Do a double pass so we can take ownership in the second one.
+                if values
+                    .iter()
+                    .all(|(key, _value)| key.parse::<usize>().is_ok())
+                {
+                    let numeric_iter: BTreeMap<_, _> = values
+                        .into_iter()
+                        .map(|(key, value)| {
+                            (key.parse::<usize>().expect("Already validated"), value)
+                        })
+                        .collect();
+                    SeqDeserializer::new(converter(self.config, numeric_iter))
+                        .deserialize_seq(visitor)
+                } else {
+                    // Just go through them in the order we found them
+                    SeqDeserializer::new(converter(self.config, values)).deserialize_seq(visitor)
+                }
```

Comment:
> I believe this is what it already does.

Ah, I meant that it can be turned off! So that users can say `Config::sort_keys(bool)` with the default being true. 

If they want sorting -> what you implemented.
If they don't want sorting -> keep the sort order as given by the env